### ISSUE TITLE
ldap_sync_users as cronjob

### ISSUE THREAD

**Issue by BlowMaWhistle** [etianen/django-python3-ldap#167]
Hi,

I'm currently trying to make the `./manage.py ldap_sync_users` a cronjob within a docker container so the application im builing gets synced every x hours. 

My problem however is that the command wont execute through the cronjob, also I'm well aware this might not be the proper place to ask since the issue lies most likely within cron but I don't know where else to ask. 

If I run any other `./manage.py` command like `migrate` or whatsoever it gives me output and tries to actually do stuff however on `ldap_sync_users` nothing is happening.

If I run the command by hand within the container everything works like a charm so the problem is within cron I guess.

Anyone have experience with this?

Regards,

**Comment by busla** [comment]
can you post your code?

I would also like to recommend not using cronjob but a scheduler of some sort (Celery beat, RQ, etc).

**Comment by BlowMaWhistle** [comment]
what part of code do you wish to see? the actual cronjob?

is celery beat able to execute ./manage.py commands? 

thanks for the quick reply tho

**Comment by etianen** [comment]
A cronjob is absolutely fine for running regular management commands, IMO.
Unless you need to coordinate work across multiple servers, or have
depenent jobs, it's overkill. Cron is *made* for running periodic scripts!

Running managerment commands in a crontab typically requires specifying the
full path to the python executable and the script. Something like:

/usr/bin/python /opt/yourapp/manage.py ldap_sync_users

**Comment by busla** [comment]
Yes, your Dockerfile, more specifically, the cronscript. Like etianen says, it's perfectly fine to run a cronjob in the container.

My point is, setting up a scheduler right from the start and writing tasks instead of a cronjob will setup the boilerplate code to run background tasks in general (send emails, cleanups, etc.). I rarely deploy projects without needing to run background tasks.

**Comment by BlowMaWhistle** [comment]
my crontab file looks as following:

`PATH=/usr/local/pyenv/shims:/usr/local/pyenv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin`
`*  *  *  *  * root python /opt/application/manage.py ldap_sync_users &>> /var/log/cron.log`
`*  *  *  *  * root python /opt/application/manage.py migrate >> /var/log/cron.log`

if i add these lines, the `./manage.py migrate` executes successfully however the `./manage.py ldap_sync_users` does nothing at all. no error messages just nothing, not even a stacktrace or something

i have verified that cron adds and executes the cronjobs from the crontab successfully its just the struggle with `ldap_sync_users`

thanks again for the quick reply to both of you

**Comment by etianen** [comment]
Huh. I'm afraid that I have absolutely no idea why ldap_sync_users might
behave differently in cron vs run manually.

If you get to the bottom of this, please let us know what the
problem/solution was!

(It's a good point about getting a "real" scheduler in early rather than
later. I guess it's just cost/benefit based on easy setup with cron, vs
time to port over to something else if requirements outgrow cron. The
tradeoff point will depend heavily on the sort of project you're writing
regularly. My projects exist in a space where celery is either far too
much, or far too little!)

**Comment by busla** [comment]
@BlowMaWhistle you didn't mention in your post that you have already tried to run `ldap_sync_users` manually. Does that work?

**Comment by busla** [comment]
>If I run the command by hand within the container everything works like a charm so the problem is within cron I guess.

sry, missed that.

**Comment by BlowMaWhistle** [comment]
Ye if I run it manually everything works perfect, as soon as I move it to cron the trouble starts

**Comment by BlowMaWhistle** [comment]
I finally made some progress. there was error output, it was just not being redirected right.
`&>>` doesn't work within crontab, maybe because its `sh` instead of `bash`? 

by correctly redirecting output with `1> /var/log/cron.log 2> /var/log/cron.err` I finally got the error message `LDAPOperationsErrorResult` now its just a matter of supplying cron with the correct ENV vars. 

thanks for the help guys! this issue can be closed :)
